GVRf/Framework/src/org/gearvrf/GVRContext.java

```diff
@@ -628,118 +634,144 @@ public AiNode wrapSceneNode(Object parent, Object matrix,
+                    final String AMBIENT_COLOR_KEY = "ambient_color";
+                    final String COLOR_OPACITY_KEY = "opacity";
+
+                    final String VERTEX_SHADER = "attribute vec4 a_position;\n"
+                            + "uniform mat4 u_mvp;\n" + "void main() {\n"
+                            + "  gl_Position = u_mvp * a_position;\n" + "}\n";
+
+                    final String FRAGMENT_SHADER = "precision mediump float;\n"
+                            + "uniform vec4 diffuse_color;\n" //
+                            + "uniform vec4 ambient_color;\n"
+                            + "uniform float opacity;\n"
+                            + "void main() {\n" //
+                            + "  gl_FragColor = ( diffuse_color * opacity ) + ambient_color;\n"
+                            + "}\n";
+
+                    GVRCustomMaterialShaderId mShaderId;
+                    GVRMaterialMap mCustomShader = null;
+
+                    final GVRMaterialShaderManager shaderManager = this
+                            .getMaterialShaderManager();
+                    mShaderId = shaderManager.addShader(VERTEX_SHADER,
+                            FRAGMENT_SHADER);
+                    mCustomShader = shaderManager.getShaderMap(mShaderId);
+                    mCustomShader.addUniformVec4Key("diffuse_color",
+                            DIFFUSE_COLOR_KEY);
+                    mCustomShader.addUniformVec4Key("ambient_color",
+                            AMBIENT_COLOR_KEY);
+                    mCustomShader.addUniformFloatKey("opacity",
+                            COLOR_OPACITY_KEY);
+
+                    GVRMaterial meshMaterial = new GVRMaterial(this, mShaderId);
+
+                    meshMaterial.setVec4(DIFFUSE_COLOR_KEY,
+                            diffuseColor.getRed(), diffuseColor.getGreen(),
+                            diffuseColor.getBlue(), diffuseColor.getAlpha());
+
+                    meshMaterial.setVec4(AMBIENT_COLOR_KEY,
+                            ambientColor.getRed(), ambientColor.getGreen(),
+                            ambientColor.getBlue(), ambientColor.getAlpha());
+
+                    meshMaterial.setFloat(COLOR_OPACITY_KEY, opacity);
+
+                    GVRSceneObject sceneObject = new GVRSceneObject(this);
+                    GVRRenderData sceneObjectRenderData = new GVRRenderData(
+                            this);
+                    sceneObjectRenderData.setMesh(futureMesh);
+                    sceneObjectRenderData.setMaterial(meshMaterial);
+                    sceneObject.attachRenderData(sceneObjectRenderData);
+
+                    wholeSceneObject.addChildObject(sceneObject);
                 }
             }
+
+            for (int i = 0; i < node.getNumChildren(); i++) {
+                this.recurseAssimpNodes(assetRelativeFilename,
+                        wholeSceneObject, node.getChildren().get(i),
+                        wrapperProvider);
+            }
```

Comment:
Hi @chenchao1407 @deepakrawat22 ,

We have inspected the code and it looks fine.

"For example, if I have a hierarchy like man -> arms -> hands -> fingers. The hierarchy of scene objects should be like man is the direct parent of arms, arms are the parents of hands, etc.."

We think that the code that we have pushed is doing the same thing as you are saying. In fact, its the standard way of recursively loading an assimp model. Even the [assimp GitHub repo sample](https://github.com/assimp/assimp/blob/master/samples%2FSimpleTexturedOpenGL%2FSimpleTexturedOpenGL%2Fsrc%2Fmodel_loading.cpp) loads the models in a similar manner. And this [article](http://learnopengl.com/#!Model-Loading/Assimp) might clarify the reason behind this implementation.

Please have a look and if you still think that we have missed something please do let us know.

We will be happy to proceed in the suggested direction.

Thanks